### ISSUE TITLE
Improved AWS session token handling

### ISSUE THREAD

**Issue by danellecline** [mbari-org/deepsea-ai#7]
@duane-edgington can you please investigate the best practice here for handling session tokens?   
 
In the new MBARI VAA AWS sub-account, I ran into a session timeout error when training, which was confusing as I thought the training failed, but it did not, and I submitted a training job 2x; this was an expensive mistake 😞  

I'm unclear on how to deal with the timeout through this pip module.  In our previous AWS account, this was not an issue, as we didn't have a session token.  

Appreciate any help with this.

**Comment by duane-edgington** [comment]
Yes. I will research this today. 


From: "DCline" ***@***.***> 
To: "mbari-org/deepsea-ai" ***@***.***> 
Cc: "Duane Edgington" ***@***.***>, "Assign" ***@***.***> 
Sent: Monday, October 31, 2022 10:24:54 AM 
Subject: [mbari-org/deepsea-ai] Improved AWS session token handling (Issue #7) 





[ https://github.com/duane-edgington | @duane-edgington ] can you investigate the best practice here for handling session tokens? 

In the new MBARI VAA AWS sub-account, I ran into a session timeout error when training, which was confusing as I thought the training failed, but it did not, and I submitted a training job 2x; this was an expensive mistake 😞 

I'm unclear on how to deal with the timeout through this pip module. In our previous AWS account, this was not an issue, as we didn't have a session token. 

Appreciate any help with this.

**Comment by duane-edgington** [comment]
Session tokens are useful to allow temporary security access to resources. They cannot last longer than 36 hours. 

If you want to run AWS CLI commands or code inside an EC2 instance, the recommended way to get credentials is to use roles for Amazon EC2. You create an IAM role that specifies the permissions that you want to grant to applications that run on the EC2 instances. When you launch the instance, you associate the role with the instance. 

I don't think the session tokens are doing what we need. 

https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_use-resources.html 
https://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html 
https://4sysops.com/archives/iam-security-token-service-sts-temporary-security-credentials-in-aws/ 

Duane

**Comment by danellecline** [comment]
The DeepSeaAI IAM role setup with the config command ```deepsea-ai setup``` grants permission for SageMaker and EC2.  
 
Relevant code starts [here](https://github.com/mbari-org/deepsea-ai/blob/main/deepsea_ai/config/setup.py#L105). Does this need to change?

**Comment by danellecline** [comment]
@kgomes do you know why we have session tokens in this new sub account?  Is that the default? Can these be removed in control tower?

**Comment by danellecline** [comment]
Danelle,
 
The answer to this is very much going to depend on where the applications are running.
 
If you have something running on an AWS service, like EC2 or RDS, then you can associate an IAM role with the instance that has appropriate permissions. Under the hood, AWS will rotate the credentials seamlessly such that it won’t result in interruptions.
 
For the other case, using credentials on a local machine, I have to check that best practice for you.
 
Which option is of interest to you?
 
Thanks,
 
____________________
Ted Zamborsky (he/him/his)
Nonprofits Solutions Architect
[ivedw@amazon.com](mailto:ivedw@amazon.com)

**Comment by danellecline** [comment]
Hi Ted,

>> If you have something running on an AWS service, like EC2 or RDS, then you can associate an IAM role with the instance that has appropriate permissions. Under the hood, AWS will rotate the credentials seamlessly such that it won’t result in interruptions.
Interesting and impressive.  Thank you.
 
Using credentials on a local machine is how we are currently setup. This is largely because our video data exists locally so our workflows are very much centered around that access. 
 
~Danelle

**Comment by danellecline** [comment]
Hello Ted,
 
Hope that you are well and that you had a nice weekend.  
 
Is the best solution for us to simply rotate these tokens every 36 hours through some external process?   
 
~Danelle
Danelle,
 
Rotating credentials based on a timer would certainly work. I’ve also been researching a relatively new feature, called AWS IAM Anywhere, which allows you to use public key infrastructure to have local machines use IAM roles similarly to how EC2 instances can. I’ve got a blog post about it below
 
https://aws.amazon.com/blogs/security/extend-aws-iam-roles-to-workloads-outside-of-aws-with-iam-roles-anywhere/
 
The long story short is that manually rotating credentials is the quick and dirty solution whereas IAM roles anywhere will require more provisioning but likely give a better long term solution.
 
Happy to set up a. meeting to discuss if you’d like that.
 
Thanks,
 
____________________
Ted Zamborsky (he/him/his)

**Comment by danellecline** [comment]
Closing this issue for now as it appears that  IAM Anywhere is the way to solve this, which is outside the scope of this code.   Setting up IAM needs to be set up in coordination with one's organization.
